ark/utils/spatial_analysis_utils.py

```diff
@@ -266,6 +257,118 @@ def compute_close_cell_num_random(marker_nums, dist_mat, dist_lim, bootstrap_num
+        dist_mat (xarray.DataArray):
+            cells x cells matrix with the euclidian distance between centers of corresponding cells
+        dist_lim (int):
+            threshold for spatial enrichment distance proximity
+        bootstrap_num (int):
+            number of permutations
+        thresh_vec (numpy.ndarray):
+            matrix of thresholds column for markers
+        current_fov_data (pandas.DataFrame):
+            data for specific patient in expression matrix
+        current_fov_channel_data (pandas.DataFrame):
+            data of only column markers for Channel Analysis
+        cell_type_col (str):
+            the name of the column in current_fov_data which contains the FlowSOM ID
+
+    Returns:
+        numpy.ndarray:
+            Large matrix of random positive marker counts for every permutation in the bootstrap
+    """
+
+    # TODO: basic checking to see if all specified cell_type_facets
+    # exist in the FlowSOM ID col of current_fov_data
+
+    # Create close_num_rand
+    close_num_rand = np.zeros((
+        len(marker_nums, len(marker_nums, bootstrap_num), dtype='int')))
+
+    dist_mat_bin = xr.DataArray(
+        (dist_mat.values < dist_lim).astype(np.int8),
+        coords=dist_mat.coords
+    )
+
+    # we need to get a boolean array of the indices that are true for each cell type
+    cell_type_data_per_facet = {}
+
+    for cell_type in marker_facets:
+        cell_data_facet = current_fov_data[cell_type_col] == cell_type
+        cell_type_data_per_facet[str(cell_type)] = cell_data_facet
+
+    # the else column will contain all the cell types not specified in cell_type_facets
+    cell_type_data_per_facet['else'] = ~current_fov_data[cell_type_col].isin(cell_type_facets)
+
+    # this is where things get a bit repetitive, however, if we wish to separate close_num
+    # and close_num_random_context we may end up having to do this, then again
+    # this is non-optimized, we can probably look into doing something more sophisticated
+    # in the future
+    num = len(thresh_vec)
+
+    for j in range(num):
+        # we need to regenerate the positive inds per marker so we can compare them
+        # with the positive indices for each cell type, needed so we can bootstrap
+        # in a context-based environment properly
+        marker1posinds = current_fov_channel_data[current_fov_channel_data.columns[j]] > thresh
+
+        # generate the number of positive hits per cell type for a specific marker
+        # in cell_type_facets or else for all the non-cell_type_facets cell types
+        cell_type_nums_per_facet_1 = {}
+        for cell_type, cell_type_data in cell_type_data_per_facet.values():
+            cell_type_nums_per_facet_1[cell_type] = np.sum(
+                np.logical_and(marker1posinds, cell_type_data.index))
+
+        # new iteration, needed to properly generate a pair of 1 vs 2 analysis
+        # basically the same thing as generating cell_type_nums_per_facet_1
```

Comment:
As it's currently written, this is exactly the same, and marker2posinds is just the first `j` entries of marker1posinds

Reply:
I'll double check Erin's code, but yeah absolutely, I think we collapse this loop out.